### ISSUE TITLE
Loading our own model for Deep Dream

### ISSUE THREAD

**Issue by amandeepsaini** [leonardvandriel/caffe2_cpp_tutorial#24]
I wanted to feed my own model into Deep Dream to see what kinds of images it would produce. I modified the code so that I could specify the init_net and predict_net to be loaded in from a local folder(similar to pretrained.cc), as opposed to specifying the model name, which would load in from the zoo keeper. 

It worked, and I was able to load in and run a locally saved squeezenet model, and also a retrained googlenet model. However, when I tried loading in a model that I had trained through different code, I ran into this issue:

> [libprotobuf FATAL /usr/include/google/protobuf/repeated_field.h:886] CHECK failed: (index) < (size()): 
> terminate called after throwing an instance of 'google::protobuf::FatalException'
>   what():  CHECK failed: (index) < (size()): 

As I'm investigating the issue further, I wanted to ask and clarify if the dream code is set up in such a way that it only supports networks that have an identical architecture to the ones in the model zoo? If not, I'm assuming that there is something at fault with the model I am loading in. Are there any guidelines that would help ensure that a model that I train would be accepted by the dream code? Thanks!

**Comment by leonardvandriel** [comment]
This error usually indicates your model was not properly loaded or is completely empty.

The dream code should work with a variety of CNN models and trainings. I added support for custom models. Instead of the model name you can provide a path with `init`/`predict` replaced by `%`. For example:

    ./bin/dream --model res/squeezenet_%_net.pb --layer fire9/concat --channel 100 --display

Let me know how that works.

**Comment by amandeepsaini** [comment]
Apologies for the delayed response, stepped away from this for a while. 

The added "%" allows me to load in my own models, yes, but the same error as above remains. So to verify whether or not the issue was originating from my own model, I tried loading in a net trained from scratch on MNIST from your tutorial code, but it also resulted in the same error. 

So currently, I can get dream to work on the models from the model zoo (googlenet & squeezenet) and if I use your code to retrain those on different datasets, but it seems that I cannot get it to work for models that are trained from scratch.

**Comment by leonardvandriel** [comment]
Interesting. Could you give me the exact commands that you run, so I can give it a try here and see if I can reproduce the issue?

**Comment by amandeepsaini** [comment]
Yup, here's an example:

> ./bin/mnist --display
**Move mnist_init_net.pb & mnist_predict_net.pb from /tmp to /res
> ./bin/dream --model res/mnist_%_net.pb --layer conv2 --channel 3 --display

**Comment by leonardvandriel** [comment]
Thanks @amandeepsaini. There's a fix in d8e6c4458. Let me know if that works.

**Comment by amandeepsaini** [comment]
Interesting.

So dream now works if I train a new MNIST from scratch using the code from your latest commit, but the one I had trained before your latest commit will not run. I see that you added the line "deploy.predict.AddOutput("softmax")" so I'm assuming that the issue originated from the way the model was created/exported. 

However, I still can't get my own models that I trained to run beyond the ones from your tutorial. For example, if I export the model from the [Python MNIST Tutorial](https://github.com/caffe2/caffe2/blob/master/caffe2/python/tutorials/MNIST.ipynb), I still can not get it to work with dream. Below is the code I added to the tutorial code to export the model:

```
def save_net(INIT_NET, PREDICT_NET, model) :
    """Save the net to the locations specified by INIT_NET and PREDICT_NET """
    with open(PREDICT_NET, 'wb') as f:
        f.write(model.net._net.SerializeToString())
    init_net = caffe2_pb2.NetDef()
    for param in model.params:
        blob = workspace.FetchBlob(param)
        shape = blob.shape
        op = core.CreateOperator("GivenTensorFill", [], [param],arg=[ utils.MakeArgument("shape", shape),utils.MakeArgument("values", blob)])
        init_net.op.extend([op])
    init_net.op.extend([core.CreateOperator("ConstantFill", [], ["data"], shape=(1,30,30))])
    with open(INIT_NET, 'wb') as f:
        f.write(init_net.SerializeToString())

def saveModel(deploy_model, modelSaveDir):
    """Save Model"""    
    with core.DeviceScope(core.DeviceOption(caffe2_pb2.CPU, 0)):

        workspace.RunNetOnce(deploy_model.param_init_net)
        workspace.CreateNet(deploy_model.net, overwrite=True)
        INIT_NET = modelSaveDir + '/init_net.pb'
        PREDICT_NET = modelSaveDir + '/predict_net.pb'
        save_net(INIT_NET, PREDICT_NET, deploy_model)

saveModel(deploy_model, 'path/to/save/dir')
```

This snippet was inserted after the "run a test pass on the test net" section, but otherwise, the tutorial code was unchanged. Any insights as to what I should modify in that tutorial or in the export code to get the exported models to work with dream? I know that the export code itself mostly works, as I've been able to load in the nets into other applications and they work just fine, but dream still gives the same error as before on them.

Thanks for all your help so far!

**Comment by leonardvandriel** [comment]
Can you share the exact error message? And can you add the output of `print(model.net.Proto())`? Thanks!

**Comment by amandeepsaini** [comment]
The error is the same as before. 

> [libprotobuf FATAL /usr/include/google/protobuf/repeated_field.h:886] CHECK failed: (index) < (size()):
terminate called after throwing an instance of 'google::protobuf::FatalException'
what(): CHECK failed: (index) < (size()):

Here is the proto:

> name: "mnist_deploy"
op {
  input: "data"
  input: "conv1_w"
  input: "conv1_b"
  output: "conv1"
  name: ""
  type: "Conv"
  arg {
    name: "kernel"
    i: 5
  }
  arg {
    name: "exhaustive_search"
    i: 0
  }
  arg {
    name: "order"
    s: "NCHW"
  }
  engine: "CUDNN"
}
op {
  input: "conv1"
  output: "pool1"
  name: ""
  type: "MaxPool"
  arg {
    name: "kernel"
    i: 2
  }
  arg {
    name: "stride"
    i: 2
  }
  arg {
    name: "cudnn_exhaustive_search"
    i: 0
  }
  arg {
    name: "order"
    s: "NCHW"
  }
  engine: "CUDNN"
}
op {
  input: "pool1"
  input: "conv2_w"
  input: "conv2_b"
  output: "conv2"
  name: ""
  type: "Conv"
  arg {
    name: "kernel"
    i: 5
  }
  arg {
    name: "exhaustive_search"
    i: 0
  }
  arg {
    name: "order"
    s: "NCHW"
  }
  engine: "CUDNN"
}
op {
  input: "conv2"
  output: "pool2"
  name: ""
  type: "MaxPool"
  arg {
    name: "kernel"
    i: 2
  }
  arg {
    name: "stride"
    i: 2
  }
  arg {
    name: "cudnn_exhaustive_search"
    i: 0
  }
  arg {
    name: "order"
    s: "NCHW"
  }
  engine: "CUDNN"
}
op {
  input: "pool2"
  input: "fc3_w"
  input: "fc3_b"
  output: "fc3"
  name: ""
  type: "FC"
  arg {
    name: "use_cudnn"
    i: 1
  }
  arg {
    name: "order"
    s: "NCHW"
  }
  arg {
    name: "cudnn_exhaustive_search"
    i: 0
  }
}
op {
  input: "fc3"
  output: "fc3"
  name: ""
  type: "Relu"
  arg {
    name: "cudnn_exhaustive_search"
    i: 0
  }
  arg {
    name: "order"
    s: "NCHW"
  }
  engine: "CUDNN"
}
op {
  input: "fc3"
  input: "pred_w"
  input: "pred_b"
  output: "pred"
  name: ""
  type: "FC"
  arg {
    name: "use_cudnn"
    i: 1
  }
  arg {
    name: "order"
    s: "NCHW"
  }
  arg {
    name: "cudnn_exhaustive_search"
    i: 0
  }
}
op {
  input: "pred"
  output: "softmax"
  name: ""
  type: "Softmax"
  arg {
    name: "order"
    s: "NCHW"
  }
  arg {
    name: "cudnn_exhaustive_search"
    i: 0
  }
  engine: "CUDNN"
}
external_input: "data"
external_input: "conv1_w"
external_input: "conv1_b"
external_input: "conv2_w"
external_input: "conv2_b"
external_input: "fc3_w"
external_input: "fc3_b"
external_input: "pred_w"
external_input: "pred_b"

**Comment by amandeepsaini** [comment]
Yup, that's it! Now if I specify the external output, I can get other networks I trained on my own to work with the dream code as well. Thanks for all your work on this, its a very handy tool! :)
